components/feature/toolbar/src/main/java/mozilla/components/feature/toolbar/ToolbarPresenter.kt

```diff
@@ -42,48 +43,77 @@ class ToolbarPresenter(
+            flow.ifAnyChanged { arrayOf(it.findCustomTabOrSelectedTab(customTabId), it.extensions) }
                 .collect { state ->
-                    if (state == null) {
-                        clear()
-                    } else {
-                        render(state)
-                    }
+                    render(state)
                 }
         }
     }
 
     fun stop() {
         scope?.cancel()
-
         renderer.stop()
     }
 
     @VisibleForTesting(otherwise = PRIVATE)
-    internal fun render(tab: SessionState) {
-        renderer.post(tab.content.url)
+    internal fun render(state: BrowserState) {
+        val tab = state.findCustomTabOrSelectedTab(customTabId)
+
+        if (tab != null) {
+            renderer.post(tab.content.url)
+
+            toolbar.setSearchTerms(tab.content.searchTerms)
+            toolbar.displayProgress(tab.content.progress)
+
+            toolbar.siteSecure = if (tab.content.securityInfo.secure) {
+                Toolbar.SiteSecurity.SECURE
+            } else {
+                Toolbar.SiteSecurity.INSECURE
+            }
+
+            toolbar.siteTrackingProtection = when {
+                tab.trackingProtection.ignoredOnTrackingProtection -> SiteTrackingProtection.OFF_FOR_A_SITE
+                tab.trackingProtection.enabled && tab.trackingProtection.blockedTrackers.isNotEmpty() ->
+                    SiteTrackingProtection.ON_TRACKERS_BLOCKED
+
+                tab.trackingProtection.enabled -> SiteTrackingProtection.ON_NO_TRACKERS_BLOCKED
 
-        toolbar.setSearchTerms(tab.content.searchTerms)
-        toolbar.displayProgress(tab.content.progress)
+                else -> SiteTrackingProtection.OFF_GLOBALLY
+            }
 
-        toolbar.siteSecure = if (tab.content.securityInfo.secure) {
-            Toolbar.SiteSecurity.SECURE
+            renderWebExtensionActions(state, tab)
         } else {
-            Toolbar.SiteSecurity.INSECURE
+            clear()
         }
+    }
 
-        toolbar.siteTrackingProtection = when {
-            tab.trackingProtection.ignoredOnTrackingProtection -> SiteTrackingProtection.OFF_FOR_A_SITE
-            tab.trackingProtection.enabled && tab.trackingProtection.blockedTrackers.isNotEmpty() ->
-                SiteTrackingProtection.ON_TRACKERS_BLOCKED
-
-            tab.trackingProtection.enabled -> SiteTrackingProtection.ON_NO_TRACKERS_BLOCKED
+    @VisibleForTesting(otherwise = PRIVATE)
+    internal fun renderWebExtensionActions(state: BrowserState, tab: SessionState) {
+        val extensions = if (tab.extensions.isNotEmpty()) {
+            tab.extensions
+        } else {
+            state.extensions
+        }
 
-            else -> SiteTrackingProtection.OFF_GLOBALLY
+        extensions.forEach { extension ->
+            extension.browserAction?.let { extensionAction ->
+                val toolbarAction = convertToToolbarAction(extensionAction)
+                toolbar.addBrowserAction(toolbarAction)
+            }
```

Comment:
I think we'll have to call `invalidateActions` on the BrowserToolbar after this so it redraws?

Reply:
Good catch! 
Unfortunately, `invalidateActions` is not part of the [Toolbar interface](https://github.com/mozilla-mobile/android-components/blob/master/components/concept/toolbar/src/main/java/mozilla/components/concept/toolbar/Toolbar.kt) is just defined on [BrowserToolbar](https://github.com/mozilla-mobile/android-components/blob/708f41ac193939316037dbbab3750d74095a181d/components/browser/toolbar/src/main/java/mozilla/components/browser/toolbar/BrowserToolbar.kt#L583). We can't have access to it without doing an explicit cast `(toolbar as BrowserToolbar).invalidateActions()` and It won't be a good idea to introduce concrete classes. I'm not sure if I should added it as @pocmo is doing some drastic modifications on the toolbar  on https://github.com/mozilla-mobile/android-components/issues/1514 and could cause some headaches for him. What do you thing @pocmo?

Reply:
Another thing we have to consider here is that the SessionState can update multiple times. So, right now we'd be adding the BrowserAction multiple times as well.

Reply:
Yeap, for this reason I filed https://github.com/mozilla-mobile/android-components/issues/4651. I will add comment pointing to this issue.

Reply:
I think that right now we are re-rendering the whole toolbar every time we have changes on `BrowserAction`s, as we are observing changes on both `tabs` or `extensions`.

```kotlin
    fun start() {
        renderer.start()

        scope = store.flowScoped { flow ->
            flow.ifAnyChanged { arrayOf(it.findCustomTabOrSelectedTab(customTabId), it.extensions) }
                .collect { state ->
                    render(state)
                }
        }
    }
```

Reply:
I think adding `invalidateActions()` to the `Toolbar` interface makes sense. It already knows about adding actions, so it should know about invalidating them too.

Reply:
I will include it as part of this pr